### DISCUSSION TITLE
New plugin, how to refresh the session token?

### DISCUSSION THREAD

**Discussion by ManuLinares** [streamlink/streamlink#6215] [Q&A] [answered]
### Description

I'm writing a new plugin for a site.

1) http.get the site for a valid token
2) Return the HLSStream.parse_variant_playlist with the correct url and token.
3) the stream starts
4) The token expires... and streamlink exits

How can I intercept the `[stream.hls][warning] No new segments in playlist for more than 6.00s. Stopping...` in the plugin and refetch a token and pass a new correct url and token to HLSStream.parse_variant_playlist?

----

I'll leave the code I'm working on the current plugin

```python
import logging
import re
import json

from streamlink.plugin import Plugin, pluginmatcher
from streamlink.plugin.api import useragents, validate
from streamlink.stream.hls import HLSStream

log = logging.getLogger(__name__)

# Define the URL matching pattern for Website
@pluginmatcher(pattern=re.compile(r"https?://(?:www\.)?website\.com/sub/[^/]+"))
class Website(Plugin):
    def _get_streams(self):
        """Extract streams from the Website webpage."""
        
        # Set the User-Agent header
        headers = {
            "User-Agent": useragents.FIREFOX,
            "Accept": "application/json; charset=utf-8",
            "Accept-Language": "en-US,en;q=0.5",
            "Content-Type": "application/json; charset=utf-8",
            "Origin": "https://www.website.com",
            "DNT": "1",
            "Connection": "keep-alive",
            "Referer": "https://www.website.com/",
            "user-country": "NL"
        }

        # Fetch the HTML page and extract the `twitter:image` meta tag content
        data = self._fetch_page_data(headers)
        if not data:
            return

        some_id = self._extract_some_id(data)
        if not some_id:
            return

        access_token, manifest = self._fetch_access_token_and_manifest(some_id, headers)
        if not access_token or not manifest:
            return

        m3u8_url = self._fetch_m3u8_url(manifest, access_token)
        if m3u8_url:
            return HLSStream.parse_variant_playlist(self.session, m3u8_url)

    def _fetch_page_data(self, headers):
        """Fetch the page data and return the meta tag content."""
        return self.session.http.get(
            self.url, 
            headers=headers, 
            schema=validate.Schema(
                validate.parse_html(),  
                validate.xml_xpath_string(".//meta[@name='twitter:image'][1]/@content"),
            )
        )

    def _extract_some_id(self, data):
        """Extract some ID from the given data."""
        match = re.search(r'/galerias/(\d+)/', data)
        if match:
            some_id = match.group(1)
            log.debug(f"Extracted some_id: {some_id}")
            return some_id
        log.error("Could not extract the some_id from the meta tag content.")
        return None

    def _fetch_access_token_and_manifest(self, some_id, headers):
        """Fetch access token and manifest from the API."""
        api_url = "https://apiv2.website.com/apiendpoint/someapi"
        payload = json.dumps({
            "id": some_id, 
            "mode": "getnewtoken", 
        })

        log.debug("Getting access token and manifest.")
        response = self.session.http.post(api_url, data=payload, headers=headers)
        response_data = self.session.http.json(response)

        error_message = response_data.get("error")
        if error_message:
            log.error(f"Error: {error_message}, Code: {response_data.get('code')}.")
            return None, None

        access_token = response_data.get("message", {}).get("accessToken")
        manifest = response_data.get("message", {}).get("manifest")
        
        log.debug(f"AccessToken: {access_token}, Manifest: {manifest}")
        return access_token, manifest

    def _fetch_m3u8_url(self, manifest, access_token):
        """Fetch the m3u8 URL from the manifest."""
        manifest_url = f"{manifest}?=&accessToken={access_token}&img=true&vdc=true&sbp=true"
        log.debug(f"Fetching manifest from: {manifest_url}")

        final_response = self.session.http.get(manifest_url, headers={
            "Host": "invertred-prod-manifest2.generflow.com",
            "Accept": "*/*",
            "User-Agent": useragents.FIREFOX,
            "Connection": "keep-alive"
        })

        final_response_json = final_response.json()  
        m3u8_url = final_response_json.get("formats", {}).get("mp4-hls", {}).get("manifest")

        if m3u8_url:
            log.debug(f"Extracted m3u8 URL: {m3u8_url}")
            return m3u8_url
        log.error("Failed to extract m3u8 URL from the manifest response.")
        return None

__plugin__ = Website
```

**Comment by bastimeyer** [comment, marked as answer]
> refetch a token and pass a new correct url and token to HLSStream.parse_variant_playlist

`HLSStream.parse_variant_playlist()` fetches and parses a **multivatiant** playlist **once**, and then returns one or more **media** playlists referenced by that multivariant playlist. Expirations of multivariant playlists are irrelevant, because a plugin will access it immediately when fetching streams. Expiring media playlists on the other hand require an override of their `url` attribute, because this is what will be polled by the worker thread continuously. Expirations are not part of the HLS specification and are always dependent on how the server implements them, so there's no generic logic for that on the client-side.

`HLSStream(HTTPStream).url` is a `@property`, so you can subclass `HLSStream` and override it with custom expiration logic:
https://streamlink.github.io/api/stream.html#streamlink.stream.HTTPStream.url

Example:
- https://github.com/streamlink/streamlink/blob/6.10.0/src/streamlink/plugins/filmon.py#L44-L91
- https://github.com/streamlink/streamlink/blob/6.10.0/tests/plugins/test_filmon.py#L43-L79

Be aware though that Streamlink doesn't define any stable APIs here other than the `HLSStream` constructor or the `HLSStream.parse_variant_playlist()` classmethod.

**Comment by ManuLinares** [comment]
I don't really understand. in the _get_streams I should return this for example?

```python
return TokenRefreshingHLSStream.parse_variant_playlist(self.session, m3u8_url)
```

Then create a class that inherits HLSStream?

```python
class TokenRefreshingHLSStream(HLSStream):
    def __init__(self, session, url: str, **kwargs):
        HLSStream.__init__(self, session, url, **kwargs)
        self.m3u8_url = url

    @property
    def url(self) -> str:
        # Check if the token has expired (e.g., by checking the time elapsed since the last token refresh)
        # If expired, refetch the token and update the URL
        if self._token_has_expired():
            self.m3u8_url = self._refetch_token_and_url()
            return self.m3u8_url
        return self.m3u8_url

    def _token_has_expired(self):
        # Token expires every 10 seconds
        ...

    def _refetch_token_and_url(self):
        #Get new token url
        ...
        return new_token_url
```

and that should continually expire the url every 10 seconds, and keep playing?

**Comment by davdn1** [comment]
Hello.
@bastimeyer Is it possible to do the same thing (renew an URL/Token, after x time) with a DASH stream?
I tried to find plugins to use as a template, but I only found for HLS streams.
Thank you.

**Reply by bastimeyer** [reply]
It's currently not possible without making significant hacks of non-public interfaces of Streamlink's DASH implementation in your custom plugins. These relevant parts of the DASH implementation are already planned to be changed in the future. See #6346 and its parent + sibling issues. I will keep manifest URL updates in mind.

**Reply by davdn1** [reply]
@bastimeyer Thank you for your answer and your insights.
I won't ask for an ETA of the proposed rewrite of the DASH implementation, because I know it's a very complex task and can take a long time, but I will keep an eye on future developments, because this kind of time limited urls are getting used more often, so having a way to update them easily would be fantastic.
I would also like take the chance to thank you @bastimeyer and other people involved in improving and maintaining Streamlink.
